File: 108. Convert Sorted Array to Binary Search Tree/tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    TreeNode* sortedArrayToBST(vector<int>& v) {
        if(v.size() == 0)
            return nullptr;
        
        return build(v, 0, v.size() - 1);
    }
    
    TreeNode* build(vector<int>& v, int low, int high){
        if(low > high)
            return nullptr;
        if(low == high)
            return new TreeNode(v[low]);
        
        int mid = (low + high)/2;
        TreeNode* left = build(v, low, mid - 1);
        TreeNode* right = build(v, mid + 1, high);
        TreeNode* root = new TreeNode(v[mid]);
        root->left = left;
        root->right = right;
        
        return root;
    }
};

File: offer/JZ26 树的子结构/tree.java
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
public class Solution {
    public Boolean isMatching(TreeNode root, TreeNode node){
        if(node == null)
            return true;
        if(root == null || root.val != node.val)
            return false;
        return isMatching(root.left, node.left) && isMatching(root.right, node.right);
    }
    public boolean HasSubtree(TreeNode root1,TreeNode root2) {
        if(root1 == null ||root2 == null)
            return false;
        if(isMatching(root1, root2))
            return true;
        else
            return HasSubtree(root1.left, root2) || HasSubtree(root1.right, root2);
    }
}